File: src/main/java/com/refinedmods/refinedstorage/item/blockitem/PortableGridBlockItem.java
package com.refinedmods.refinedstorage.item.blockitem;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.RSBlocks;
import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.apiimpl.network.grid.factory.PortableGridGridFactory;
import com.refinedmods.refinedstorage.inventory.player.PlayerSlot;
import com.refinedmods.refinedstorage.item.WirelessGridItem;
import com.refinedmods.refinedstorage.render.Styles;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;

import javax.annotation.Nullable;
import java.util.List;

public class PortableGridBlockItem extends EnergyBlockItem {
    private final Type type;

    public PortableGridBlockItem(Type type) {
        super(
            type == Type.CREATIVE ? RSBlocks.CREATIVE_PORTABLE_GRID.get() : RSBlocks.PORTABLE_GRID.get(),
            new Item.Properties().stacksTo(1),
            type == Type.CREATIVE,
            () -> RS.SERVER_CONFIG.getPortableGrid().getCapacity()
        );

        this.type = type;
    }

    public Type getType() {
        return type;
    }

    @Override
    public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand hand) {
        ItemStack stack = player.getItemInHand(hand);

        if (!level.isClientSide) {
            API.instance().getGridManager().openGrid(PortableGridGridFactory.ID, (ServerPlayer) player, stack, PlayerSlot.getSlotForHand(player, hand));
        }

        return InteractionResultHolder.success(stack);
    }

    @Override
    public void appendHoverText(ItemStack stack, @Nullable Level level, List<Component> tooltip, TooltipFlag flag) {
        super.appendHoverText(stack, level, tooltip, flag);

        tooltip.add(Component.translatable("block.refinedstorage.portable_grid.tooltip").setStyle(Styles.GRAY));
    }

    @Override
    public InteractionResult useOn(UseOnContext context) {
        if (context.getPlayer() == null) {
            return InteractionResult.FAIL;
        }

        //Place
        if (context.getPlayer().isCrouching()) {
            return super.useOn(context);
        }

        ItemStack stack = context.getPlayer().getItemInHand(context.getHand());

        if (!context.getLevel().isClientSide) {
            API.instance().getGridManager().openGrid(PortableGridGridFactory.ID, (ServerPlayer) context.getPlayer(), stack, PlayerSlot.getSlotForHand(context.getPlayer(), context.getHand()));
        }

        return InteractionResult.SUCCESS;
    }

    @Override
    public int getEntityLifespan(ItemStack stack, Level level) {
        return Integer.MAX_VALUE;
    }

    @Override
    public boolean shouldCauseReequipAnimation(ItemStack oldStack, ItemStack newStack, boolean slotChanged) {
        if (oldStack.getItem() == newStack.getItem() &&
            WirelessGridItem.getSortingDirection(oldStack) == WirelessGridItem.getSortingDirection(newStack) &&
            WirelessGridItem.getSortingType(oldStack) == WirelessGridItem.getSortingType(newStack) &&
            WirelessGridItem.getSearchBoxMode(oldStack) == WirelessGridItem.getSearchBoxMode(newStack) &&
            WirelessGridItem.getTabSelected(oldStack) == WirelessGridItem.getTabSelected(newStack) &&
            WirelessGridItem.getTabPage(oldStack) == WirelessGridItem.getTabPage(newStack) &&
            WirelessGridItem.getSize(oldStack) == WirelessGridItem.getSize(newStack)) {
            return false;
        }

        return super.shouldCauseReequipAnimation(oldStack, newStack, slotChanged);
    }

    public enum Type {
        NORMAL,
        CREATIVE
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/blockitem/PortableGridBlockItem.kt
package com.refinedmods.refinedstorage.item.blockitem

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.apiimpl.API.Companion.instance
import com.refinedmods.refinedstorage.apiimpl.network.grid.factory.PortableGridGridFactory
import com.refinedmods.refinedstorage.item.WirelessGridItem
import com.refinedmods.refinedstorage.render.Styles
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.world.World
import java.util.function.Supplier

class PortableGridBlockItem(val type: Type) : EnergyBlockItem(
        if (type == Type.CREATIVE) BlockRegistryGenerated.CREATIVE_PORTABLE_GRID else BlockRegistryGenerated.PORTABLE_GRID,
        Settings().group(RS.MAIN_GROUP).maxCount(1),
        type == Type.CREATIVE,
        Supplier { RS.SERVER_CONFIG.portableGrid.getCapacity() }
) {
    enum class Type {
        NORMAL, CREATIVE
    }

    fun onItemRightClick(world: World, player: PlayerEntity, hand: Hand?): ActionResult<ItemStack> {
        val stack: ItemStack = player.getHeldItem(hand)
        if (!world.isClient) {
            instance().getGridManager()!!.openGrid(PortableGridGridFactory.ID, player as ServerPlayerEntity, stack, player.inventory.currentItem)
        }
        return ActionResult.resultSuccess(stack)
    }

    override fun addInformation(stack: ItemStack, @Nullable world: World?, tooltip: MutableList<Text?>, flag: ITooltipFlag?) {
        super.addInformation(stack, world, tooltip, flag)
        tooltip.add(TranslationTextComponent("block.refinedstorage.portable_grid.tooltip").setStyle(Styles.GRAY))
    }

    fun onItemUse(context: ItemUseContext): ActionResult {
        return if (!context.getPlayer().isCrouching()) {
            ActionResult.FAIL
        } else super.onItemUse(context)
    }

    fun getEntityLifespan(stack: ItemStack?, world: World?): Int {
        return Int.MAX_VALUE
    }

    fun shouldCauseReequipAnimation(oldStack: ItemStack, newStack: ItemStack, slotChanged: Boolean): Boolean {
        if (oldStack.item === newStack.item) {
            if (WirelessGridItem.Companion.getSortingDirection(oldStack) == WirelessGridItem.Companion.getSortingDirection(newStack) && WirelessGridItem.Companion.getSortingType(oldStack) == WirelessGridItem.Companion.getSortingType(newStack) && WirelessGridItem.Companion.getSearchBoxMode(oldStack) == WirelessGridItem.Companion.getSearchBoxMode(newStack) && WirelessGridItem.Companion.getTabSelected(oldStack) == WirelessGridItem.Companion.getTabSelected(newStack) && WirelessGridItem.Companion.getTabPage(oldStack) == WirelessGridItem.Companion.getTabPage(newStack) && WirelessGridItem.Companion.getSize(oldStack) == WirelessGridItem.Companion.getSize(newStack)) {
                return false
            }
        }
        return super.shouldCauseReequipAnimation(oldStack, newStack, slotChanged)
    }

    init {
        this.setRegistryName(RS.ID, (if (type == Type.CREATIVE) "creative_" else "") + "portable_grid")
    }
}